gossipsub.go

```diff
@@ -0,0 +1,613 @@
+		mesh:    make(map[string]map[peer.ID]struct{}),
+		fanout:  make(map[string]map[peer.ID]struct{}),
+		lastpub: make(map[string]int64),
+		gossip:  make(map[peer.ID][]*pb.ControlIHave),
+		control: make(map[peer.ID]*pb.ControlMessage),
+		mcache:  NewMessageCache(GossipSubHistoryGossip, GossipSubHistoryLength),
+	}
+	return NewPubSub(ctx, h, rt, opts...)
+}
+
+// GossipSubRouter is a router that implements the gossipsub protocol.
+// For each topic we have joined, we maintain an overlay through which
+// messages flow; this is the mesh map.
+// For each topic we publish to without joining, we maintain a list of peers
+// to use for injecting our messages in the overlay with stable routes; this
+// is the fanout map. Fanout peer lists are expired if we don't publish any
+// messages to their topic for GossipSubFanoutTTL.
+type GossipSubRouter struct {
+	p       *PubSub
+	peers   map[peer.ID]protocol.ID         // peer protocols
+	mesh    map[string]map[peer.ID]struct{} // topic meshes
+	fanout  map[string]map[peer.ID]struct{} // topic fanout
+	lastpub map[string]int64                // last pubish time for fanout topics
+	gossip  map[peer.ID][]*pb.ControlIHave  // pending gossip
+	control map[peer.ID]*pb.ControlMessage  // pending control messages
+	mcache  *MessageCache
+}
+
+func (gs *GossipSubRouter) Protocols() []protocol.ID {
+	return []protocol.ID{GossipSubID, FloodSubID}
+}
+
+func (gs *GossipSubRouter) Attach(p *PubSub) {
+	gs.p = p
+	go gs.heartbeatTimer()
+}
+
+func (gs *GossipSubRouter) AddPeer(p peer.ID, proto protocol.ID) {
+	gs.peers[p] = proto
+}
+
+func (gs *GossipSubRouter) RemovePeer(p peer.ID) {
+	delete(gs.peers, p)
+	for _, peers := range gs.mesh {
+		delete(peers, p)
+	}
+	for _, peers := range gs.fanout {
+		delete(peers, p)
+	}
+	delete(gs.gossip, p)
+	delete(gs.control, p)
+}
+
+func (gs *GossipSubRouter) HandleRPC(rpc *RPC) {
+	ctl := rpc.GetControl()
+	if ctl == nil {
+		return
+	}
+
+	iwant := gs.handleIHave(ctl)
+	ihave := gs.handleIWant(ctl)
+	prune := gs.handleGraft(rpc.from, ctl)
+	gs.handlePrune(rpc.from, ctl)
+
+	if len(iwant) == 0 && len(ihave) == 0 && len(prune) == 0 {
+		return
+	}
+
+	out := rpcWithControl(ihave, nil, iwant, nil, prune)
+	gs.sendRPC(rpc.from, out)
+}
+
+func (gs *GossipSubRouter) handleIHave(ctl *pb.ControlMessage) []*pb.ControlIWant {
+	iwant := make(map[string]struct{})
+
+	for _, ihave := range ctl.GetIhave() {
+		topic := ihave.GetTopicID()
+		_, ok := gs.mesh[topic]
+		if !ok {
+			continue
+		}
+
+		for _, mid := range ihave.GetMessageIDs() {
+			if gs.p.seenMessage(mid) {
+				continue
+			}
+			iwant[mid] = struct{}{}
+		}
+	}
+
+	if len(iwant) == 0 {
+		return nil
+	}
+
+	iwantlst := make([]string, 0, len(iwant))
+	for mid := range iwant {
+		iwantlst = append(iwantlst, mid)
+	}
+
+	return []*pb.ControlIWant{&pb.ControlIWant{MessageIDs: iwantlst}}
+}
+
+func (gs *GossipSubRouter) handleIWant(ctl *pb.ControlMessage) []*pb.Message {
+	ihave := make(map[string]*pb.Message)
```

Comment:
Does this need to be a map, or could it be a slice? Is it a map to deduplicate message IDs?

Reply:
yes, it needs to deduplicate.